{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "from omnipath._core.downloader._downloader import Downloader\n",
    "import omnipath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "import liana as li"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "dnwld = Downloader()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show homologues available\n",
    "raw_taxa_url = \"https://raw.githubusercontent.com/oganm/homologene/master/data-raw/taxData.tsv\"\n",
    "homlogene_url = \"https://raw.githubusercontent.com/oganm/homologene/master/data-raw/homologene2.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_homologene():\n",
    "    return dnwld.maybe_download(raw_taxa_url, callback=pd.read_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>name_txt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28985</td>\n",
       "      <td>Kluyveromyces lactis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>318829</td>\n",
       "      <td>Magnaporthe oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>33169</td>\n",
       "      <td>Eremothecium gossypii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4896</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4932</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5141</td>\n",
       "      <td>Neurospora crassa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7165</td>\n",
       "      <td>Anopheles gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>8364</td>\n",
       "      <td>Xenopus (Silurana) tropicalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>9031</td>\n",
       "      <td>Gallus gallus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>9544</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>9598</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>9615</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>9913</td>\n",
       "      <td>Bos taurus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tax_id                       name_txt\n",
       "0    10090                   Mus musculus\n",
       "1    10116              Rattus norvegicus\n",
       "2    28985           Kluyveromyces lactis\n",
       "3   318829             Magnaporthe oryzae\n",
       "4    33169          Eremothecium gossypii\n",
       "5     3702           Arabidopsis thaliana\n",
       "6     4530                   Oryza sativa\n",
       "7     4896      Schizosaccharomyces pombe\n",
       "8     4932       Saccharomyces cerevisiae\n",
       "9     5141              Neurospora crassa\n",
       "10    6239         Caenorhabditis elegans\n",
       "11    7165              Anopheles gambiae\n",
       "12    7227        Drosophila melanogaster\n",
       "13    7955                    Danio rerio\n",
       "14    8364  Xenopus (Silurana) tropicalis\n",
       "15    9031                  Gallus gallus\n",
       "16    9544                 Macaca mulatta\n",
       "17    9598                Pan troglodytes\n",
       "18    9606                   Homo sapiens\n",
       "19    9615         Canis lupus familiaris\n",
       "20    9913                     Bos taurus"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show_homologene()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "# omnipath.clear_cache()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Homologene Download function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_homologene_raw(): \n",
    "    homologene = (dnwl.maybe_download(homlogene_url,\n",
    "                                     callback=pd.read_table,\n",
    "                                     ).\n",
    "                  astype(str)\n",
    "                  )\n",
    "    return homologene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "homologene = _get_homologene_raw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_homologene(source_organism, target_organism):\n",
    "    \n",
    "    homologene = _get_homologene_raw().set_index(\"HID\")\n",
    "    \n",
    "    source_df = homologene[(homologene[['Taxonomy']] == source_organism).values][['Gene.Symbol']]\n",
    "    target_df = homologene[(homologene[['Taxonomy']] == target_organism).values][['Gene.Symbol']]\n",
    "\n",
    "    homologene = pd.merge(source_df, target_df,\n",
    "                          right_index=True, left_index=True,\n",
    "                          suffixes=('_source', '_target'),\n",
    "                          how='inner')\n",
    "    homologene = homologene.reset_index().rename({f'Gene.Symbol_source':'source', f'Gene.Symbol_target':'target'}, axis=1)\n",
    "    homologene = homologene[['source', 'target']]\n",
    "    \n",
    "    return homologene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Homology Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "map_df = download_homologene('9606', '10090')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ETFA</td>\n",
       "      <td>Etfa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CLDN4</td>\n",
       "      <td>Cldn4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HAUS2</td>\n",
       "      <td>Haus2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EFHC1</td>\n",
       "      <td>Efhc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LRRC8D</td>\n",
       "      <td>Lrrc8d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17307</th>\n",
       "      <td>ODAD2</td>\n",
       "      <td>Odad2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17308</th>\n",
       "      <td>SRBD1</td>\n",
       "      <td>Srbd1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17309</th>\n",
       "      <td>WRAP53</td>\n",
       "      <td>Wrap53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17310</th>\n",
       "      <td>ZNF358</td>\n",
       "      <td>Zfp358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17311</th>\n",
       "      <td>KRTAP16-1</td>\n",
       "      <td>Krtap16-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17312 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          source     target\n",
       "0           ETFA       Etfa\n",
       "1          CLDN4      Cldn4\n",
       "2          HAUS2      Haus2\n",
       "3          EFHC1      Efhc1\n",
       "4         LRRC8D     Lrrc8d\n",
       "...          ...        ...\n",
       "17307      ODAD2      Odad2\n",
       "17308      SRBD1      Srbd1\n",
       "17309     WRAP53     Wrap53\n",
       "17310     ZNF358     Zfp358\n",
       "17311  KRTAP16-1  Krtap16-1\n",
       "\n",
       "[17312 rows x 2 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "resource = li.resource.select_resource()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "resource = li.resource.explode_complexes(resource)\n",
    "# map_df = dict(zip(map_df['source'], map_df['target']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _rebuild_complexes(df, group_cols, target_col):\n",
    "    df = resource.copy()\n",
    "    df_grouped = resource.groupby(group_cols)['target'].agg(lambda x: '_'.join(map(str, x))).reset_index()\n",
    "    df = df.drop('target', axis=1).merge(df_grouped, on=list(np.setdiff1d(group_cols, target_col)), how='inner')\n",
    "    df = df.drop(f'{target_col}_complex', axis=1).rename({'target':f'{target_col}_complex'}, axis=1)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction</th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "      <th>ligand_complex</th>\n",
       "      <th>receptor_complex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LGALS9&amp;PTPRC</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>PTPRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LGALS9&amp;MET</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>MET</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>MET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LGALS9&amp;CD44</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>CD44</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>CD44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LGALS9&amp;LRP1</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>LRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LGALS9&amp;CD47</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>CD47</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>CD47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5849</th>\n",
       "      <td>BMP2&amp;ACTR2</td>\n",
       "      <td>BMP2</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>BMP2</td>\n",
       "      <td>ACTR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5850</th>\n",
       "      <td>BMP15&amp;ACTR2</td>\n",
       "      <td>BMP15</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>BMP15</td>\n",
       "      <td>ACTR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5851</th>\n",
       "      <td>CSF1&amp;CSF3R</td>\n",
       "      <td>CSF1</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>CSF1</td>\n",
       "      <td>CSF3R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5852</th>\n",
       "      <td>IL36G&amp;IFNAR1</td>\n",
       "      <td>IL36G</td>\n",
       "      <td>IFNAR1</td>\n",
       "      <td>IL36G</td>\n",
       "      <td>IFNAR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5853</th>\n",
       "      <td>IL36G&amp;IFNAR2</td>\n",
       "      <td>IL36G</td>\n",
       "      <td>IFNAR2</td>\n",
       "      <td>IL36G</td>\n",
       "      <td>IFNAR2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5854 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       interaction  ligand receptor ligand_complex receptor_complex\n",
       "0     LGALS9&PTPRC  LGALS9    PTPRC         LGALS9            PTPRC\n",
       "1       LGALS9&MET  LGALS9      MET         LGALS9              MET\n",
       "2      LGALS9&CD44  LGALS9     CD44         LGALS9             CD44\n",
       "3      LGALS9&LRP1  LGALS9     LRP1         LGALS9             LRP1\n",
       "4      LGALS9&CD47  LGALS9     CD47         LGALS9             CD47\n",
       "...            ...     ...      ...            ...              ...\n",
       "5849    BMP2&ACTR2    BMP2    ACTR2           BMP2            ACTR2\n",
       "5850   BMP15&ACTR2   BMP15    ACTR2          BMP15            ACTR2\n",
       "5851    CSF1&CSF3R    CSF1    CSF3R           CSF1            CSF3R\n",
       "5852  IL36G&IFNAR1   IL36G   IFNAR1          IL36G           IFNAR1\n",
       "5853  IL36G&IFNAR2   IL36G   IFNAR2          IL36G           IFNAR2\n",
       "\n",
       "[5854 rows x 5 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resource"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "resource = resource.merge(map_df, left_on='receptor', right_on='source', how='inner').drop(['source'], axis=1)\n",
    "resource = _rebuild_complexes(resource, ['interaction', 'ligand'], 'receptor')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction</th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "      <th>ligand_complex</th>\n",
       "      <th>receptor_complex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LGALS9&amp;PTPRC</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>Ptprc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD22&amp;PTPRC</td>\n",
       "      <td>CD22</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>CD22</td>\n",
       "      <td>Ptprc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRC1&amp;PTPRC</td>\n",
       "      <td>MRC1</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>MRC1</td>\n",
       "      <td>Ptprc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LGALS1&amp;PTPRC</td>\n",
       "      <td>LGALS1</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>LGALS1</td>\n",
       "      <td>Ptprc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LGALS9&amp;MET</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>MET</td>\n",
       "      <td>LGALS9</td>\n",
       "      <td>Met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5720</th>\n",
       "      <td>GDF1&amp;ACTR2</td>\n",
       "      <td>GDF1</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>GDF1</td>\n",
       "      <td>Actr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5721</th>\n",
       "      <td>INHBA&amp;ACTR2</td>\n",
       "      <td>INHBA</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>INHBA</td>\n",
       "      <td>Actr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5722</th>\n",
       "      <td>BMP7&amp;ACTR2</td>\n",
       "      <td>BMP7</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>BMP7</td>\n",
       "      <td>Actr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5723</th>\n",
       "      <td>BMP2&amp;ACTR2</td>\n",
       "      <td>BMP2</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>BMP2</td>\n",
       "      <td>Actr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5724</th>\n",
       "      <td>BMP15&amp;ACTR2</td>\n",
       "      <td>BMP15</td>\n",
       "      <td>ACTR2</td>\n",
       "      <td>BMP15</td>\n",
       "      <td>Actr2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5725 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       interaction  ligand receptor ligand_complex receptor_complex\n",
       "0     LGALS9&PTPRC  LGALS9    PTPRC         LGALS9            Ptprc\n",
       "1       CD22&PTPRC    CD22    PTPRC           CD22            Ptprc\n",
       "2       MRC1&PTPRC    MRC1    PTPRC           MRC1            Ptprc\n",
       "3     LGALS1&PTPRC  LGALS1    PTPRC         LGALS1            Ptprc\n",
       "4       LGALS9&MET  LGALS9      MET         LGALS9              Met\n",
       "...            ...     ...      ...            ...              ...\n",
       "5720    GDF1&ACTR2    GDF1    ACTR2           GDF1            Actr2\n",
       "5721   INHBA&ACTR2   INHBA    ACTR2          INHBA            Actr2\n",
       "5722    BMP7&ACTR2    BMP7    ACTR2           BMP7            Actr2\n",
       "5723    BMP2&ACTR2    BMP2    ACTR2           BMP2            Actr2\n",
       "5724   BMP15&ACTR2   BMP15    ACTR2          BMP15            Actr2\n",
       "\n",
       "[5725 rows x 5 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resource"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "resource = resource.merge(map_df, left_on='ligand', right_on='source', how='inner').drop(['source'], axis=1)\n",
    "resource = _rebuild_complexes(resource, ['interaction', 'receptor'], 'ligand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction</th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "      <th>receptor_complex</th>\n",
       "      <th>ligand_complex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CD22&amp;PTPRC</td>\n",
       "      <td>CD22</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>Ptprc</td>\n",
       "      <td>Cd22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MRC1&amp;PTPRC</td>\n",
       "      <td>MRC1</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>Ptprc</td>\n",
       "      <td>Mrc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LGALS1&amp;PTPRC</td>\n",
       "      <td>LGALS1</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>Ptprc</td>\n",
       "      <td>Lgals1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LGALS1&amp;ITGB1</td>\n",
       "      <td>LGALS1</td>\n",
       "      <td>ITGB1</td>\n",
       "      <td>Itgb1</td>\n",
       "      <td>Lgals1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LGALS1&amp;CD69</td>\n",
       "      <td>LGALS1</td>\n",
       "      <td>CD69</td>\n",
       "      <td>Cd69</td>\n",
       "      <td>Lgals1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5427</th>\n",
       "      <td>SERPINA7&amp;SLC16A2</td>\n",
       "      <td>SERPINA7</td>\n",
       "      <td>SLC16A2</td>\n",
       "      <td>Slc16a2</td>\n",
       "      <td>Serpina7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428</th>\n",
       "      <td>BTN1A1&amp;TARM1</td>\n",
       "      <td>BTN1A1</td>\n",
       "      <td>TARM1</td>\n",
       "      <td>Tarm1</td>\n",
       "      <td>Btn1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5429</th>\n",
       "      <td>BTN1A1&amp;NEGR1</td>\n",
       "      <td>BTN1A1</td>\n",
       "      <td>NEGR1</td>\n",
       "      <td>Negr1</td>\n",
       "      <td>Btn1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5430</th>\n",
       "      <td>IGFBP7&amp;MPIG6B</td>\n",
       "      <td>IGFBP7</td>\n",
       "      <td>MPIG6B</td>\n",
       "      <td>Mpig6b</td>\n",
       "      <td>Igfbp7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5431</th>\n",
       "      <td>BPI&amp;SLAMF9</td>\n",
       "      <td>BPI</td>\n",
       "      <td>SLAMF9</td>\n",
       "      <td>Slamf9</td>\n",
       "      <td>Bpi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5432 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           interaction    ligand receptor receptor_complex ligand_complex\n",
       "0           CD22&PTPRC      CD22    PTPRC            Ptprc           Cd22\n",
       "1           MRC1&PTPRC      MRC1    PTPRC            Ptprc           Mrc1\n",
       "2         LGALS1&PTPRC    LGALS1    PTPRC            Ptprc         Lgals1\n",
       "3         LGALS1&ITGB1    LGALS1    ITGB1            Itgb1         Lgals1\n",
       "4          LGALS1&CD69    LGALS1     CD69             Cd69         Lgals1\n",
       "...                ...       ...      ...              ...            ...\n",
       "5427  SERPINA7&SLC16A2  SERPINA7  SLC16A2          Slc16a2       Serpina7\n",
       "5428      BTN1A1&TARM1    BTN1A1    TARM1            Tarm1         Btn1a1\n",
       "5429      BTN1A1&NEGR1    BTN1A1    NEGR1            Negr1         Btn1a1\n",
       "5430     IGFBP7&MPIG6B    IGFBP7   MPIG6B           Mpig6b         Igfbp7\n",
       "5431        BPI&SLAMF9       BPI   SLAMF9           Slamf9            Bpi\n",
       "\n",
       "[5432 rows x 5 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resource"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove duplicates (1:many mappings)\n",
    "resource = resource.drop_duplicates(['interaction']).drop(['interaction'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "      <th>receptor_complex</th>\n",
       "      <th>ligand_complex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CD22</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>Ptprc</td>\n",
       "      <td>Cd22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MRC1</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>Ptprc</td>\n",
       "      <td>Mrc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LGALS1</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>Ptprc</td>\n",
       "      <td>Lgals1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LGALS1</td>\n",
       "      <td>ITGB1</td>\n",
       "      <td>Itgb1</td>\n",
       "      <td>Lgals1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LGALS1</td>\n",
       "      <td>CD69</td>\n",
       "      <td>Cd69</td>\n",
       "      <td>Lgals1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5427</th>\n",
       "      <td>SERPINA7</td>\n",
       "      <td>SLC16A2</td>\n",
       "      <td>Slc16a2</td>\n",
       "      <td>Serpina7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428</th>\n",
       "      <td>BTN1A1</td>\n",
       "      <td>TARM1</td>\n",
       "      <td>Tarm1</td>\n",
       "      <td>Btn1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5429</th>\n",
       "      <td>BTN1A1</td>\n",
       "      <td>NEGR1</td>\n",
       "      <td>Negr1</td>\n",
       "      <td>Btn1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5430</th>\n",
       "      <td>IGFBP7</td>\n",
       "      <td>MPIG6B</td>\n",
       "      <td>Mpig6b</td>\n",
       "      <td>Igfbp7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5431</th>\n",
       "      <td>BPI</td>\n",
       "      <td>SLAMF9</td>\n",
       "      <td>Slamf9</td>\n",
       "      <td>Bpi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4034 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ligand receptor receptor_complex ligand_complex\n",
       "0         CD22    PTPRC            Ptprc           Cd22\n",
       "1         MRC1    PTPRC            Ptprc           Mrc1\n",
       "2       LGALS1    PTPRC            Ptprc         Lgals1\n",
       "3       LGALS1    ITGB1            Itgb1         Lgals1\n",
       "4       LGALS1     CD69             Cd69         Lgals1\n",
       "...        ...      ...              ...            ...\n",
       "5427  SERPINA7  SLC16A2          Slc16a2       Serpina7\n",
       "5428    BTN1A1    TARM1            Tarm1         Btn1a1\n",
       "5429    BTN1A1    NEGR1            Negr1         Btn1a1\n",
       "5430    IGFBP7   MPIG6B           Mpig6b         Igfbp7\n",
       "5431       BPI   SLAMF9           Slamf9            Bpi\n",
       "\n",
       "[4034 rows x 4 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resource"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spiana",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b688e7f183144fde965166e58483e32fc6dbc6fc380ccc51d49ef608da0385a8"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
